src/librustc_mir/borrow_check.rs

```diff
@@ -487,7 +487,22 @@ impl<'c, 'b, 'a: 'b+'c, 'gcx, 'tcx: 'a> MirBorrowckCtxt<'c, 'b, 'a, 'gcx, 'tcx>
             if flow_state.inits.curr_state.contains(&mpi) {
                 // may already be assigned before reaching this statement;
                 // report error.
-                self.report_illegal_reassignment(context, (lvalue, span));
+                let mut assigned_lvalue = &move_data.move_paths[mpi].lvalue;
+                loop {
+                    match *assigned_lvalue {
+                        Lvalue::Projection(ref proj) => {
+                            assigned_lvalue = &proj.base;
+                            continue;
+                        }
+                        Lvalue::Local(local) => {
+                            let assigned_span = self.mir.local_decls[local].source_info.span;
```

Comment:
Does the source info for the declaration always correspond to the assigned span when reported by AST borrowck?

In particular, I'm thinking of cases like:

```rust
let declare_without_init: i32;
match result {
    Ok(_) => declare_without_init = 1,
    Err(_) => declare_without_init = 2,
}
...
declare_without_init = 3;
```

where the AST-borrowck today, for better or worse, tags the assignment in the first match arm as the "first assignment"

Reply:
I don't actually know, I've only scoured the source code and used my own deduction skills to do what makes sense in order to retrieve a span. Perhaps @nikomatsakis may know more about this?

Reply:
I just compiled the example you gave with my patch -- it actually does the wrongest thing, which is displaying the span of `let declare_without_init: i32;`. It would seem like `move_path_to_lvalue` would not provide the location in which a variable is assigned.